api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,209 @@
+     * Once a particular event is connected ot a particular patient, this tracks this change for the patient
+     * @param patientId the patient id related to the event
+     * @param event the event to track
+     */
+    public void handlePatientEvent(Integer patientId, DbEvent event) {
+        if (log.isDebugEnabled()) {
+            log.debug("Patient Event: (patient_id=" + patientId + ") - " + event);
+        }
+        Timestamp lastUpdated = new Timestamp(event.getTimestamp());
+        String sql = "insert into dbevent_patient (patient_id, last_updated) values (?, ?) on duplicate key update last_updated = ?";
+        database.executeUpdate(sql, patientId, lastUpdated, lastUpdated);
+        if (event.getTable().equals("patient")) {
+            if (event.getOperation() == Operation.DELETE || event.getValues().getBoolean("voided")) {
+                sql = "update dbevent_patient set deleted = true where patient_id = ? and deleted = false";
+                database.executeUpdate(sql, patientId);
+            }
+            else if (!event.getValues().getBoolean("voided")) {
+                sql = "update dbevent_patient set deleted = false where patient_id = ? and deleted = true";
+                database.executeUpdate(sql, patientId);
+            }
+        }
+    }
+
+    /**
+     * Rather than stream all initial READ events one by one, it is much more efficient to set up the initial state
+     * by querying the database directly to record the most recent date updated and deleted status for all patients.
+     * This method performs this operation, and returns the max last_updated in the database
+     */
+    protected synchronized void performInitialSnapshot() {
+        if (!snapshotInitialized) {
+            log.warn("Performing initial snapshot into dbevent_patient from patient");
+            database.executeUpdate(
+                    "insert ignore into dbevent_patient (patient_id, last_updated, deleted) " +
+                            "select patient_id, greatest(date_created, ifnull(date_changed, date_created), ifnull(date_voided, date_created)), voided from patient"
+            );
+            for (DatabaseTable table : config.getMonitoredTables()) {
+                String tableName = table.getTableName();
+                if (config.isIncluded(table) && !tableName.equals("patient")) {
+                    Set<String> columns = table.getColumns().keySet();
+
+                    List<String> dateCols = new ArrayList<>();
+                    if (columns.contains("date_created")) {
+                        dateCols.add("ifnull(t.date_created, p.last_updated)");
+                    }
+                    if (columns.contains("date_changed")) {
+                        dateCols.add("ifnull(t.date_changed, p.last_updated)");
+                    }
+                    if (columns.contains("date_voided")) {
+                        dateCols.add("ifnull(t.date_voided, p.last_updated)");
+                    }
```

Comment:
If we identify other columns that we want to factor in, we can add them here.  I wasn't immediately aware of any.